Source/AlamofireExtended.swift

```diff
@@ -0,0 +1,55 @@
+//
+//  AlamofireExtended.swift
+//
+//  Copyright (c) 2019 Alamofire Software Foundation (http://alamofire.org/)
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+/// Type that acts as a generic extension point for all `AlamofireExtended` types.
+public struct AlamofireExtension<ExtendedType> {
+    /// Stores the type or metatype of any extended type.
+    let type: ExtendedType
+
+    init(_ type: ExtendedType) {
+        self.type = type
+    }
+}
+
+/// Protocol describing the `af` extension points for Alamofire extended types.
+public protocol AlamofireExtended {
+    associatedtype ExtendedType
+
+    /// Static Alamofire extension point.
+    static var af: AlamofireExtension<ExtendedType>.Type { get set }
+    /// Instance Alamofire extension point.
+    var af: AlamofireExtension<ExtendedType> { get set }
+}
+
+public extension AlamofireExtended {
+    static var af: AlamofireExtension<Self>.Type {
+        get { return AlamofireExtension<Self>.self }
+        set { }
+    }
+
+    var af: AlamofireExtension<Self> {
+        get { return AlamofireExtension(self) }
+        set { }
```

Comment:
I've never built a no-op `set` implementation before. Makes sense that it can just be empty. I wonder if we could just switch these to this:

```swift
static var af: AlamofireExtension<Self>.Type { return AlamofireExtension<Self>.self }
var af: AlamofireExtension<Self> { return AlamofireExtension(self) }
```

Reply:
I copied the [implementation from RxSwift](https://github.com/ReactiveX/RxSwift/blob/master/RxSwift/Reactive.swift), except for the comment in the `set`, which states: "// this enables using Reactive to "mutate" base type". So while I'm not sure we need it, it may be useful.

Reply:
I wanted to learn more about this so I put together a simple playground and learned a few things. First off, this is a cool pattern for sure, very powerful. Secondly, there are a couple hidden tricks that aren't super obvious at first.

### 1) Setting only works for reference types

The `ExtendedType` in the `AlamofireExtension` is a `let`, so you can only mutate the state of the `ExtendedType` instance if it's a reference type. You could switch it to a `var`, but that causes more harm than good.

Also, you MUST have the `set` declaration in `AlamofireExtended` for being able to modify state on the `ExtendedType` instance. Otherwise you get the following error:

> "Cannot assign to property: 'af' is a get-only property"

### 2) Value types can use these APIs, but "setting" state doesn't work as expected

Since a new `AlamofireExtension` type is made on each `af` call, the internal type is copied if it's a value type. Therefore, setting things on one instance won't probably end up what you want.

```swift
import Foundation
import Security

public struct AlamofireExtension<ExtendedType> {
    let type: ExtendedType
    init(_ type: ExtendedType) { self.type = type }
}

public protocol AlamofireExtended {
    associatedtype ExtendedType

    static var af: AlamofireExtension<ExtendedType>.Type { get set }
    var af: AlamofireExtension<ExtendedType> { get set }
}

public extension AlamofireExtended {
    static var af: AlamofireExtension<Self>.Type {
        get { return AlamofireExtension<Self>.self }
        set { }
    }

    var af: AlamofireExtension<Self> {
        get { return AlamofireExtension(self) }
        set { print("Called with new value: \(newValue)") }
    }
}

extension URLRequest: AlamofireExtended {}
public extension AlamofireExtension where ExtendedType == URLRequest {
    var method: String {
        get { return type.httpMethod ?? "GET" }
//        set { type.httpMethod = newValue }
    }
}

extension URLSession: AlamofireExtended {}
extension AlamofireExtension where ExtendedType: URLSession {
    var fancyDescription: String {
        get { return type.sessionDescription ?? "" }
        set { type.sessionDescription = newValue }
    }
}

var request = URLRequest(url: URL(string: "https://httpbin.org/post")!)
request.httpMethod = "POST"

let method = request.af.method
//print("before set")
//request.af.method = "HEAD"
//print("after set")

let method2 = request.af.method

var session = URLSession(configuration: .default)
let desc = session.af.fancyDescription
print("before set")
session.af.fancyDescription = "This is cool stuff!"
print("after set")
let desc2 = session.af.fancyDescription
```